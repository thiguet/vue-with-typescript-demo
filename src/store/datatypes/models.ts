interface Unique {
    id?: string;
}

export enum Measures {
    UN = 'UN',
    L = 'L',
    KG = 'KG',
    G = 'G',
}

export interface Product extends Unique {
    name: string;
    measure: Measures;
    qtd: number;
    minQtd: number;
    image?: string;
}

export interface User extends Unique {
    name: string;
    email: string;
}

export interface ReportItem {
    text: string;
    value: number;
}

export type ReportData = ReportItem[];

export interface QuantityReportData {
    productName: string;
    productQuantity: number;
}

export type QuantityReport = QuantityReportData[];
export interface MeasuresReportData {
    measureName: string;
    measureQuantity: number;
}

export type MeasuresReport = MeasuresReportData[];

export enum VuexAppModules {
    products = 'products',
    login = 'login',
    alert = 'alert',
    reports = 'reports',
}
