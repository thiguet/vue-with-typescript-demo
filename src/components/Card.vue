<template>
    <div class="container" @click="onclick">
        <img :src="icon" />
        <slot />
    </div>
</template>

<script lang="ts">
import Vue from 'vue';

import Component from 'vue-class-component';
import { Prop } from 'vue-property-decorator';

@Component
export default class Card extends Vue {
    @Prop({ required: true }) readonly onclick!: Function;

    @Prop({ required: true }) readonly icon!: string;
}
</script>

<style>
.container {
    cursor: pointer;
    justify-content: center;
    padding: 0px;
    background-color: var(--ternary-bg);
    height: 70vh;
    width: 300px;
}

.container > img {
    max-width: 150px;
    max-height: 150px;
    align-self: center;
}

@media screen and (max-width: 600px) {
    .container {
        height: 32vh;
        padding: 10px;
        width: 85%;
        justify-content: center;
        align-self: center;
        margin-top: 20px;
    }

    .container > img {
        max-width: 120px;
        max-height: 120px;
    }
}
</style>
