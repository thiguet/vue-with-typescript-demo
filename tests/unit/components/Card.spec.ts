import { shallowMount, mount, createLocalVue, Wrapper } from '@vue/test-utils';

import Vuex from 'vuex';

import Card from '@/components/Card.vue';
import faker from 'faker';
import { CardProps } from './models';

const localVue = createLocalVue();
localVue.use(Vuex);

describe('Login', () => {
    const slotContent = '<div>Main Content</div>';
    let props: CardProps;

    const build = () => {
        const componentBindings = {
            propsData: { ...props },
            slots: {
                default: slotContent,
            },
        };

        const wrapper: Wrapper<Card> = shallowMount(Card, componentBindings);

        const mountedWrapper: Wrapper<Card> = mount(Card, componentBindings);

        const CardComp = () => {
            const CardInstance = (wrapper.vm as unknown) as CardProps;

            jest.spyOn(CardInstance, 'onclick');

            return CardInstance;
        };

        return {
            wrapper,
            mountedWrapper,
            CardComp,
            container: () => wrapper.find('div'),
            img: () => wrapper.find('img'),
        };
    };

    beforeEach(() => {
        props = {
            onclick: jest.fn(),
            icon: faker.image.imageUrl(),
        };
    });

    it('renders component', () => {
        const { wrapper } = build();
        expect(wrapper).toMatchSnapshot();
    });

    it('renders child components', () => {
        const { container } = build();
        expect(container().exists()).toBe(true);
    });

    it('renders slot data', () => {
        const { wrapper } = build();
        expect(wrapper.html()).toContain(slotContent);
    });

    it('renders slot data', async () => {
        const { wrapper, CardComp } = build();
        await wrapper.trigger('click');
        expect(CardComp().onclick).toHaveBeenCalled();
    });

    it('sets icon prop to the src of the img component', async () => {
        const { img } = build();
        expect(img().attributes('src')).toBe(props.icon);
    });
});
