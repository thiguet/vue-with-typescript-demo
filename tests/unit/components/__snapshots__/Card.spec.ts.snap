// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders component 1`] = `
<div class="container"><img src="http://lorempixel.com/640/480">
  <div>Main Content</div>
</div>
`;
